doc/rm/asm_coding_style.md

```diff
@@ -0,0 +1,286 @@
+
+```S
+  // Correct:
+  lw t3, 8(sp)
+  sb t3, 0(a0)
+  // Wrong:
+  lw t3, sp, 8
+  sb t3, a0
+```
+
+### Compressed Instruction Mnemonics
+
+***Do not use compressed instruction mnemonics.***
+
+While Ibex implements the RISC-V C extension, it is expected that the toolchain will automatically compress instructions where possible.
+This advice extends to the `c.unimp` all-zeroes illegal instruction.
+Toolchains like `LLVM` and `gas` are expected to recognize the `unimp` mnemonic to stand for two `c.unimp` instructions.
+
+Of course, this advice should be ignored when it is necessary to prove that a certain block of instructions does not exceed a particular width.
+
+### "Current Point" Label
+
+***Do not use the current point (`.`) label.***
+
+The current point label does not look like a label, and can be easilly missed during review.
+
+## `.S` Files
+
+This advice applies specifically to `.S` files, as well as globally-scoped assembly in `.c` and `.cc` files.
+
+While this is is already implicit, we only use the `.S` extension for assembly files; not `.s` or `.asm`.
+
+### Indentation
+
+***Assembly files must be formatted with all directives indented two spaces, except for labels. Comments should be indented as usual.***
+
+Example:
+```S
+_trap_start:
+  .globl _trap_start
+  csrr a0, mcause
+  // ...
+```
+
+### Comments
+
+***Comments must use either the `//` or `/* */` syntaxes.***
+
+Every function-like label which is meant to be called like a function (*especially* `.globl`s) should be given a Doxygen-style comment.
+While Doxygen is not suited for assembly, that style should be used for consistency.
+See the [C/C++ style guide]({{< relref "c_cpp_coding_style" >}}) for more information.
+
+All other advice for writing comments, as in the C/C++ style guide, also applies.
+
+### Register useage
+
+***Register usage in a "function" that diverges from the RISC-V function call ABI must be documented.***
+
+This includes non-standard calling conventions, non-standard clobbers, and other behavior not expected of a well-behaved RISV-V function.
+Non-standard input and output registers should use Doxygemn's `param[in] reg` and
+`param[out] reg` annotations, respectively.
+
+Example:
+```S
+/**
+ * Compute some stuff, outputing a 96-bit integer.
+ *
+ * @param[out] a0 bits [31:0] of the result.
+ * @param[out] a1 bits [63:32] of the result.
+ * @param[out] a2 bits [95:64] of the result.
+ */
+compute_stuff:
+  .globl compute_stuff
+  li a0, 0xdeadbeef
+  li a1, 0xbeefcafe
+  li a2, 0xcafedead
+  ret
+```
+
+### Ending an Instruction Sequence
+
+***Every code path within an assembly file must end in a non-linking jump.***
+
+Assembly should be written such that the program counter can't wander off past the written instructions.
+As such, all assembly should be ended with `ret` (or any of the protection ring returns like `mret`), an infinite `wfi` loop, or an instruction that is guaranteed to trap and not return, like an `exit`-like syscall.
```

Comment:
If we're happy with `unimp` here, probably worth listing explicitly?

Reply:
Done.